README.md
# **Electric Ammo**

---

Adds electric ammos, with power level similar to piercing rounds, that can be obtained from charged batteries of [Freight Forwarding](https://mods.factorio.com/mod/FreightForwarding) or [Battery Powered](https://mods.factorio.com/mod/battery-powered)

## Features

- Adds a new technology between military and chemical science to unlock Electric Ammo
- Get Electric Ammos from charged batteries' electric source
- Transferring energy between the objects has a small % chance of failing, but you can recharge empty batteries again

![showcase-electric-ammo](https://github.com/RedRafe/electric-ammo/blob/main/archive/showcase.png?raw=true)

## Translation
Available languages:
- English (en)

You can help by translating this mod into your language. Any new translations will be included in the next release.

---

## Known Bugs & Mod compatibility
Please feel free to report any known issue! :) 

- Requires [Freight Forwarding](https://mods.factorio.com/mod/FreightForwarding) or [Battery Powered](https://mods.factorio.com/mod/battery-powered) mod
- Also supports [Krastorio2](https://mods.factorio.com/mod/Krastorio2) ammos & graphics

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: 2023.04.25
  Bugfixes:
    - Literally nothing to see here O:)
---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 2023.04.25
  Changes:
    - Guess what? Another compatibility update
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2023.04.24
  Info:
    - Initial release


prototypes/item.lua
-- Electric Rounds Magazine 

data:extend({
  -- Ammo
  {
    type = "ammo",
    name = "electric-rounds-magazine",
    icon = "__electric-ammo__/graphics/electric-rounds-magazine.png",
    icon_size = 64, icon_mipmaps = 4,
    ammo_type = {
      category = "bullet",
      action =
      {
        type = "direct",
        action_delivery =
        {
          type = "instant",
          source_effects =
          {
            type = "create-explosion",
            entity_name = "explosion-gunshot"
          },
          target_effects =
          {
            {
              type = "create-entity",
              entity_name = "explosion-hit",
              offsets = {{0, 1}},
              offset_deviation = {{-0.5, -0.5}, {0.5, 0.5}}
            },
            {
              type = "damage",
              damage = { amount = 8, type = "physical"}
            }
          }
        }
      }
    },
    magazine_size = 10,
    subgroup = "ammo",
    order = "a[basic-clips]-b[piercing-rounds-magazine]-2",
    stack_size = 200
  },
  -- Recipe
  {
    type = "recipe",
    name = "electric-rounds-magazine",
    main_product = "electric-rounds-magazine",
    category = "advanced-crafting",
    energy_required = 3,
    ingredients = {{ElectricAmmo.BATTERY_NAME, 2}},
    results = {
      {
        name = "battery",
        probability = 0.95,
        amount = 2
      },
      {
        name = "electric-rounds-magazine",
        probability = 0.95,
        amount = 1
      }
    },
    emissions_multiplier = 8,
    enabled = false,
    allow_as_intermediate = false,
    always_show_made_in = true,
    show_amount_in_title = false,
    always_show_products = true,
  },
})

prototypes/electric-rounds-magazine.lua
-- Electric Rounds Magazine 

data:extend({
  -- Ammo
  {
    type = "ammo",
    name = "electric-rounds-magazine",
    icon = "__electric-ammo__/graphics/electric-rounds-magazine.png",
    icon_size = 64, icon_mipmaps = 4,
    ammo_type = {
      category = "bullet",
      action =
      {
        type = "direct",
        action_delivery =
        {
          type = "instant",
          source_effects =
          {
            type = "create-explosion",
            entity_name = "explosion-gunshot"
          },
          target_effects =
          {
            {
              type = "create-entity",
              entity_name = "explosion-hit",
              offsets = {{0, 1}},
              offset_deviation = {{-0.5, -0.5}, {0.5, 0.5}}
            },
            {
              type = "damage",
              damage = { amount = 8, type = "physical"}
            }
          }
        }
      }
    },
    magazine_size = 10,
    subgroup = "ammo",
    order = "a[basic-clips]-b[piercing-rounds-magazine]-2",
    stack_size = 200
  },
  -- Recipe
  {
    type = "recipe",
    name = "electric-rounds-magazine",
    main_product = "electric-rounds-magazine",
    category = "advanced-crafting",
    energy_required = 3,
    ingredients = {{"ff-charged-battery", 2}},
    results = {
      {
        name = "battery",
        probability = 0.95,
        amount = 2
      },
      {
        name = "electric-rounds-magazine",
        probability = 0.95,
        amount = 1
      }
    },
    emissions_multiplier = 8,
    enabled = false,
    allow_as_intermediate = false,
    always_show_made_in = true,
    show_amount_in_title = false,
    always_show_products = true,
  },
})

prototypes/krastorio2.lua
if not mods["Krastorio2"] then return end

data:extend({
  -- Pistol ammo
  {
    type = "ammo",
    name = "electric-pistol-rounds-magazine",
    icon = "__electric-ammo__/graphics/kr-pistol-ammo.png",
    icon_size = 64, icon_mipmaps = 4,
    pictures = {
      {
        filename = "__electric-ammo__/graphics/kr-pistol-ammo.png",
        mipmap_count = 4,
        scale = 0.25,
        size = 64
      }
    },
    ammo_type = {
      action = {
        {
          action_delivery = {
            {
              direction_deviation = 0.15,
              max_range = 22,
              projectile = "pistol-ammo-2",
              range_deviation = 0.15,
              source_effects = {
                {
                  entity_name = "explosion-gunshot",
                  type = "create-explosion"
                }
              },
              starting_speed = 1.1,
              type = "projectile"
            }
          },
          type = "direct"
        }
      },
      category = "pistol-ammo",
      target_type = "direction"
    },
    localised_name = {
      "item-name.kr-pistol-ammo"
    },
    magazine_size = 10,
    subgroup = "ammo",
    order = "a[basic-clips]-a02[pistol-magazine]-2",
    stack_size = 200,
  },
  -- Rifle ammo
  {
    type = "ammo",
    name = "electric-rounds-magazine",
    icon = "__electric-ammo__/graphics/kr-rifle-ammo.png",
    icon_size = 64, icon_mipmaps = 4,
    pictures = {
      {
        filename = "__electric-ammo__/graphics/kr-rifle-ammo.png",
        mipmap_count = 4,
        scale = 0.25,
        size = 64
      }
    },
    ammo_type = {
      action = {
        {
          action_delivery = {
            {
              direction_deviation = 0.15,
              max_range = 30,
              projectile = "rifle-ammo-2",
              range_deviation = 0.15,
              source_effects = {
                {
                  entity_name = "explosion-gunshot",
                  type = "create-explosion"
                }
              },
              starting_speed = 1.75,
              type = "projectile"
            }
          },
          type = "direct"
        }
      },
      category = "bullet",
      target_type = "direction"
    },
    localised_name = {
      "item-name.kr-rifle-ammo"
    },
    magazine_size = 30,
    subgroup = "ammo",
    order = "a[basic-clips]-a04[rifle-magazine]-2",
    stack_size = 200,
  },
  -- Anti Material ammo
  {
    type = "ammo",
    name = "electric-anti-material-rounds-magazine",
    icon = "__electric-ammo__/graphics/kr-anti-material-rifle-ammo.png",
    icon_size = 64, icon_mipmaps = 4,
    pictures = {
      {
        filename = "__electric-ammo__/graphics/kr-anti-material-rifle-ammo.png",
        mipmap_count = 4,
        scale = 0.25,
        size = 64
      }
    },
    ammo_type = {
      action = {
        {
          action_delivery = {
            {
              direction_deviation = 0.02,
              max_range = 70,
              projectile = "anti-material-rifle-ammo-2",
              range_deviation = 0.02,
              source_effects = {
                {
                  entity_name = "explosion-gunshot",
                  type = "create-explosion"
                }
              },
              starting_speed = 3,
              type = "projectile"
            }
          },
          force = "not-same",
          type = "direct"
        }
      },
      category = "anti-material-rifle-ammo",
      target_type = "direction"
    },
    localised_name = {
      "item-name.kr-anti-material-rifle-ammo"
    },
    magazine_size = 7,
    subgroup = "ammo",
    order = "a[basic-clips]-a08[anti-material-rifle-magazine]-2",
    stack_size = 200,
  },
  -- Pistol Recipe
  {
    type = "recipe",
    name = "electric-pistol-rounds-magazine",
    main_product = "electric-pistol-rounds-magazine",
    category = "advanced-crafting",
    energy_required = 3,
    ingredients = {{ElectricAmmo.BATTERY_NAME, 1}},
    results = {
      {
        name = "battery",
        probability = 0.97,
        amount = 1
      },
      {
        name = "electric-pistol-rounds-magazine",
        probability = 0.97,
        amount = 1
      }
    },
    emissions_multiplier = 8,
    enabled = false,
    allow_as_intermediate = false,
    always_show_made_in = true,
    show_amount_in_title = false,
    always_show_products = true,
  },
  -- Anti Material Recipe
  {
    type = "recipe",
    name = "electric-anti-material-rounds-magazine",
    main_product = "electric-anti-material-rounds-magazine",
    category = "advanced-crafting",
    energy_required = 3,
    ingredients = {{ElectricAmmo.BATTERY_NAME, 3}},
    results = {
      {
        name = "battery",
        probability = 0.92,
        amount = 3
      },
      {
        name = "electric-anti-material-rounds-magazine",
        probability = 0.92,
        amount = 1
      }
    },
    emissions_multiplier = 8,
    enabled = false,
    allow_as_intermediate = false,
    always_show_made_in = true,
    show_amount_in_title = false,
    always_show_products = true,
  },
})

if data.raw.technology["electric-ammo"] then 
  local effects = data.raw.technology["electric-ammo"].effects
  if effects then
    table.insert(
      effects,
      {
        type = "unlock-recipe",
        recipe = "electric-pistol-rounds-magazine"
      }
    )
    table.insert(
      effects,
      {
        type = "unlock-recipe",
        recipe = "electric-anti-material-rounds-magazine"
      }
    )
  else
    log("Electric Ammo: no effects found")
  end
else
  log("Electric Ammo: no tech found")
end